cdaweb access through requests

jklenzing opened the pull request:
Partial fix for #79.  Replaces `ftplib` with `requests` in cdaweb_methods.download to allow access to cdaweb data via HTTPS.  Allows for future implementation of restful web services with cdaweb server.

jklenzing commented:
Travis recommends switching to HTTPS.  Going to work on the updating the NOAA files, but the website is not back online yet.

https://docs.travis-ci.com/user/common-build-problems/#ftpsmtpother-protocol-does-not-work

rstoneback commented:
While you are working with requests you may also want to hit the cdaweb_methods.list_remote_files method as well.

On Jan 26, 2019, at 11:48 AM, Jeff Klenzing <notifications@github.com<mailto:notifications@github.com>> wrote:


Partial fix for #79<https://github.com/rstoneback/pysat/issues/79>. Replaces 'ftplib' with 'requests' in cdaweb_methods.download to allow access to cdaweb data via HTTPS. Allows for future implementation of restful web services with cdaweb server.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/rstoneback/pysat/pull/143

Commit Summary

  *   implemented requests in cdw.download
  *   Merge branch 'develop' into cdaweb_fix

File Changes

  *   M pysat/instruments/nasa_cdaweb_methods.py<https://github.com/rstoneback/pysat/pull/143/files#diff-0> (190)

Patch Links:

  *   https://github.com/rstoneback/pysat/pull/143.patch
  *   https://github.com/rstoneback/pysat/pull/143.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/rstoneback/pysat/pull/143>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADbded1O-MZR6OZAoprLUYzndQ0BLlpuks5vHJTpgaJpZM4aUSa3>.

jklenzing commented:
NOAA and GFZ do not seem to have HTTPS access.

jklenzing commented:
Working on the remote files issue now.  Found more description of the ftp issues here:

https://blog.travis-ci.com/2018-07-23-the-tale-of-ftp-at-travis-ci

Looks like using vpn would fix the travis errors.  Right now it only affects kp and dst.

jklenzing commented:
This one is stable and ready for review.  We should discuss the ftp options at the next meeting.

jklenzing commented:
Since this fixes a number of travis errors, I would prefer we review and merge this one first.

rstoneback commented:
Hey Jeff,

I’m planning (hoping) to take care of reviews tomorrow. Thanks for the guidance and all the commits!

Cheers,
Russell

On Jan 31, 2019, at 9:56 AM, Jeff Klenzing <notifications@github.com<mailto:notifications@github.com>> wrote:


Since this fixes a number of travis errors, I would prefer we review and merge this one first.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://github.com/rstoneback/pysat/pull/143#issuecomment-459396925>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADbdeS16axQ_bEnHmMl3xAx4XbgDgdnUks5vIxIQgaJpZM4aUSa3>.

jklenzing commented:
No worries. I was largely thinking of order of operations since the other
one will take a while to review.

On Thu, Jan 31, 2019 at 10:58 AM Russell Stoneback <notifications@github.com>
wrote:

> Hey Jeff,
>
> I’m planning (hoping) to take care of reviews tomorrow. Thanks for the
> guidance and all the commits!
>
> Cheers,
> Russell
>
> On Jan 31, 2019, at 9:56 AM, Jeff Klenzing <notifications@github.com
> <mailto:notifications@github.com>> wrote:
>
>
> Since this fixes a number of travis errors, I would prefer we review and
> merge this one first.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub<
> https://github.com/rstoneback/pysat/pull/143#issuecomment-459396925>, or
> mute the thread<
> https://github.com/notifications/unsubscribe-auth/ADbdeS16axQ_bEnHmMl3xAx4XbgDgdnUks5vIxIQgaJpZM4aUSa3>.
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstoneback/pysat/pull/143#issuecomment-459397762>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ASr0HIGSQTr6sFAVbAQleyGUm-iXE2Ihks5vIxKRgaJpZM4aUSa3>
> .
>

rstoneback commented on pysat/instruments/nasa_cdaweb_methods.py:
```diff
@@ -205,104 +209,106 @@ def download(supported_tags, date_array, tag, sat_id,
-                      ftp_site='cdaweb.gsfc.nasa.gov', 
+                print('File not available for ' + date.strftime('%x'))
+        except requests.exceptions.RequestException as exception:
+            # if exception[0][0:3] != '550':
```
These commented lines may be removed.

jklenzing replied:
fixed

rstoneback commented on pysat/instruments/omni_hro.py:
```diff
@@ -181,8 +181,8 @@ def time_shift_to_magnetic_poles(inst):
+
     return
 
 def download(date_array, tag, sat_id='', data_path=None, user=None,
```
This should probably be replaced with the CDAWeb call. I think I just missed converting OMNI when I transitioned the other instruments over.

jklenzing replied:
I had tried that initially, but was having issues with the `fake_daily_files_from_monthly` keyword.  Any suggestions?

jklenzing replied:
updated using the CDAweb call

rstoneback commented:
This build had a problem with pyglow for some reason. Other two errors related to kp and dst. Gonna restart tests for the pyglow check.

jklenzing commented:
@rstoneback, I'm still getting the following error:
```
TypeError: download() got multiple values for argument 'supported_tags'
```
Any thoughts?

jklenzing commented:
OK,found it.  Download does not treat supported_tags as a keyword, whereas list_files does.

rstoneback commented:
Clearly my reading comprehension is not so great in the morning :)

 Please ignore my last comment. supported_tags is not a keyword, as you stated.


On Feb 4, 2019, at 10:17 PM, Jeff Klenzing <notifications@github.com<mailto:notifications@github.com>> wrote:


OK,found it. Download does not treat supported_tags as a keyword, whereas list_files does.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/rstoneback/pysat/pull/143#issuecomment-460507267>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADbdedZb7PsmSVIbhRbDNUlOH9jJj6Stks5vKQW_gaJpZM4aUSa3>.

rstoneback commented:
Thanks again Jeff. I’m planning on hitting pysat reviews and other things tonight!

Cheers,
Russell

rstoneback commented:
I added a test file for ftp instruments. It should only run when on travis. I also added support to exclude particular sat_id and tag combinations from download testing.

I switched a printed warning to an actual warning (when downloads don't work). As part of this, I realized that the Dst filename standard was updated on the remote server end. Added that minor fix.

jklenzing commented:
rerunning to check timeout errors

jklenzing commented:
Looking at the logs for builds 1590 and 1592 on Travis, kp still runs and errors on 1590, dst on 1592.  There is no difference between the codes other than some whitespace.  Not sure why these are inconsistently still being tested on travis.

rstoneback commented:
I haven't sorted out why dst is being checked either. It doesn't run on my local system. Anyway, I fixed up a few things, checking to see how it runs on the server side.

rstoneback commented:
My Travis check is bad, the FTP tests are still being run.

rstoneback commented:
Remaining issues: pyglow errors, which aren't a part of this, and; supermag magnetometer indices download problem. Works locally ;) Trying things out again.

rstoneback commented:
@aburrell I had to modify supermag since it wasn't passing a more strict download test. The latest code had 'indices' files mapped to 'all' but not quite everywhere. The test thought nothing was downloaded, even though a new file was on disk. My final fix in __init__ may be the best of them, assuming I went in the correct direction.

aburrell commented:
@rstoneback that seems sensible.  It was a bit of a judgement call on organizing a subset of data that is normally only downloaded alongside a different data product but a lot of people use instead.

rstoneback commented:
Thanks! No worries about the org, just making sure I'm staying with the original intent. Looks like the 'indices' tag was sorted out, now downloads are complaining about tag='' .

aburrell commented:
@jklenzing the testing ftp testing file needs a way to add an instrument tag and not all tags of a particular instrument.

jklenzing commented:
Currently, `test_sw.py` requires downloads of FTP data as part of the setup structure.  This should be rewritten using fake data.  Conversely, we can turn these tests on in Travis CI and open new issues to expand testing for now.

jklenzing commented:
> Currently, `test_sw.py` requires downloads of FTP data as part of the setup structure. This should be rewritten using fake data. Conversely, we can turn these tests on in Travis CI and open new issues to expand testing for now.

Never mind.  self.download=False not being flagged in three tests for combine_kp for some reason.

jklenzing commented:
> @jklenzing the testing ftp testing file needs a way to add an instrument tag and not all tags of a particular instrument.

@aburrell, I agree, but I'd like to close this off as the scope has grown significantly beyond the original pull request.  Going to write up some new issues for other changes to improve the handling of instrument download testing within travis.

jklenzing commented:
One remaining Error in 3.6 as of #1678 in Travis CI:
```
ERROR: Failure: RuntimeWarning (Download for supermag magnetometer   was not successful.)
```
This fails both on Travis and locally.

2.7 still has the pyglow issues as well (fixed in #155).

jklenzing commented:
Found areas in supermag where `tag=None` was used in checks instead of `tag=''`.  Fixed these.  Unit tests now run locally.

Note: ucar_tiegcm is now popping up with errors locally.  Since there is no supported download routine, added this to exclude_list.

jklenzing commented:
removed ucar_tiegcm from exclude list.  Was an issue with how I am running nosetests locally.

jklenzing commented:
@aburrell, @rstoneback, could you take a look at supermag?  Still getting errors locally.  If I try to run and download through ipython,I get this error message:
```
~/code/pysat/pysat/instruments/supermag_magnetometer.py in append_ascii_data(file_strings, tag)
    928         lsplit = re.split('\t+', out_lines[i])
    929         dtime = pds.datetime.strptime(" ".join(lsplit[0:-1]),
--> 930                                       "%Y %m %d %H %M %S")
    931         date_list.append(dtime)
    932         num_stations.append(int(lsplit[-1]))
```
Not really sure what is going on in the append_ascii_data method.  After updating the tag structure, I was able to successfully download once for the test data, but I'm getting this error now.

aburrell commented:
@jklenzing  can you send me the code snippet that you ran to get this error?

jklenzing commented:
```
import pysat  
mag = pysat.Instrument(platform='supermag', name='magnetometer')
t0 = pysat.datetime(2009, 1, 1)     
mag.download(t0, t0, user=username)  
```
where username is my user name as a string.

rstoneback commented:
Hey Folks!

  I’ve been covered up on some other projects but there is a light at the end of the tunnel. Hopefully I can start contributing positively back here pretty soon.

I think I saw some comments about an inclusive list for ftp downloads that is tag specific. I’m planning on adding this but if someone wants to jump ahead I recently added an exclusion list that supports tags. In the FTP instrument file you can do the inverse of that solution. The current package support already there sets up the groundwork.

Cheers,
Russell



On Feb 11, 2019, at 12:02 PM, Jeff Klenzing <notifications@github.com<mailto:notifications@github.com>> wrote:


import pysat
mag = pysat.Instrument(platform='supermag', name='magnetometer')
t0 = pysat.datetime(2009, 1, 1)
mag.download(t0, t0, user=username)


where username is my user name as a string.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/rstoneback/pysat/pull/143#issuecomment-462429781>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADbdeXd39U-Ij0A4vISw2uoeqeggnrHrks5vMbAXgaJpZM4aUSa3>.

jklenzing commented:
@rstoneback, I started working on the include_tags a bit (some remnants are in this branch), but instead I'm leaning toward a proposal that I outlined in #156 as a more permanent solution for customizing the testing functions.  Comments are welcome.  We should discuss at the next meeting.

As far as what needs to be done here, If we fix (or temporarily suppress) the supermag bug, then merge this PR and #155, we should be back to a passing build on Travis.  #155 is ready for review.

aburrell commented:
> Not really sure what is going on in the append_ascii_data method. After updating the tag structure, I was able to successfully download once for the test data, but I'm getting this error now.

The error is occurring because the indices are not included when requesting just the magnetometer data, so ind_num should be zero with tag ''.

jklenzing commented:
Thanks @aburrell.  Sticking `ucar_tiegcm` back in the exclude list for now.  Not sure why this one only errors when everything else seems to be fixed.  Crossing fingers...

rstoneback commented:
I don’t think the TIEGCM code has a working download routine. All the files I’ve used it on have been sent to me by non-standard means.

If anyone knows of a data source that we can refer too we can add it in.

Cheers,
Russell

On Feb 12, 2019, at 5:24 PM, Jeff Klenzing <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks @aburrell<https://github.com/aburrell>. Sticking ucar_tiegcm back in the exclude list for now. Not sure why this one only errors when everything else seems to be fixed. Crossing fingers...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/rstoneback/pysat/pull/143#issuecomment-462982862>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADbdeeqCT9mDAkI42SfP60OdWYEMqR8Jks5vM001gaJpZM4aUSa3>.

jklenzing commented:
@rstoneback, that's right, there's currently no download support.

jklenzing commented:
Ok, this one’s ready for review.

aburrell commented:
WHAT?  All checks have passed?  :-O

aburrell commented on pysat/instruments/jro_isr.py:
```diff
@@ -93,7 +94,7 @@
 # function as the default function for (JRO).
 # Default function is run first by the nanokernel on every load call.
-default = pysat.instruments.madrigal_methods.filter_data_single_date      
+default = mad_meth.filter_data_single_date      
```
ha, made me laugh.

jklenzing replied:
Me too.  Can't take credit for that one though.

rstoneback replied:
Glad I could brighten up some days :)

aburrell commented on pysat/instruments/nasa_cdaweb_methods.py:
```diff
@@ -73,26 +73,26 @@ def list_files(tag=None, sat_id=None, data_path=None, format_str=None,
     else:
         estr = 'A directory must be passed to the loading routine for <Instrument Code>'
-        raise ValueError (estr)            
+        raise ValueError (estr)
```
Shouldn't have a space between a function and a paren (here and elsewhere)

jklenzing commented:
I’m surprised too. Wasn’t expecting that until this was combined with #155

jklenzing commented:
Hmm.  Supermag indices are now putting in an error.  Only changes post-review were changing 
```
raise ValueError (estr)
```
to 
```
raise ValueError(estr)
```
in a few places.  Restarting travis tests and running tests locally.

rstoneback commented:
I'm ok with keeping the tiegcm in the exclude list. It isn't a fully 'complete' module.

rstoneback commented:
Thanks Jeff!

On Feb 13, 2019, at 12:05 PM, Jeff Klenzing <notifications@github.com<mailto:notifications@github.com>> wrote:


Merged #143<https://github.com/rstoneback/pysat/pull/143> into develop.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/rstoneback/pysat/pull/143#event-2137505699>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADbdeZ5hdcx1oak3nX1Ek_zt0TX2IFCSks5vNFP5gaJpZM4aUSa3>.
